core-services/src/main/java/org/zalando/nakadi/service/publishing/AvroEventPublisher.java

```diff
@@ -0,0 +1,155 @@
+                            final String publishedBy,
+                            final GenericRecord event) {
+        try {
+            final GenericRecord metadata = new GenericData.Record(avroSchema.getMetadataSchema());
+            metadata.put("occurred_at", new DateTime(DateTimeZone.UTC).toString());
+            metadata.put("eid", uuidGenerator.randomUUID().toString());
+            metadata.put("flow_id", FlowIdUtils.peek());
+            metadata.put("event_type", etName);
+            metadata.put("partition", 0); // fixme avro
+            metadata.put("received_at", new DateTime(DateTimeZone.UTC));
+            metadata.put("schema_version", "0");  // fixme avro
+            metadata.put("published_by", publishedBy);
+
+            Closeable publishingCloser = null;
+            try {
+                publishingCloser = timelineSync.workWithEventType(etName, nakadiSettings.getTimelineWaitTimeoutMs());
+
+                // publish under timeline lock
+                serializeAndSendToKafka(etName, metadata, event);
+
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+                LOG.error("Failed to wait for timeline switch", e);
+            } catch (final TimeoutException e) {
+                LOG.error("Failed to wait for timeline switch", e);
+            } finally {
+                try {
+                    if (publishingCloser != null) {
+                        publishingCloser.close();
+                    }
+                } catch (final IOException e) {
+                    LOG.error("Exception occurred when releasing usage of event-type", e);
+                }
+            }
+
+        } catch (final Exception e) {
+            LOG.error("Error occurred when submitting KPI event for publishing", e);
+        }
+    }
+
+    private void serializeAndSendToKafka(final String eventTypeName,
+                                         final GenericRecord eventMetadata,
+                                         final GenericRecord event) {
+        final EventType eventType = eventTypeCache.getEventType(eventTypeName);
+        final Timeline activeTimeline = timelineService.getActiveTimeline(eventType);
+        final String topic = activeTimeline.getTopic();
+
+        final Span publishingSpan = TracingService.buildNewSpan("publishing_to_kafka")
+                .withTag(Tags.MESSAGE_BUS_DESTINATION.getKey(), topic)
+                .withTag("event_type", eventTypeName)
+                .withTag("type", "avro")
+                .start();
+
+        try (Closeable ignored = TracingService.activateSpan(publishingSpan)) {
+            final ByteArrayOutputStream eventOutputStream = new ByteArrayOutputStream();
+            final GenericDatumWriter genericDatumWriter = new GenericDatumWriter(eventMetadata.getSchema());
+            genericDatumWriter.setSchema(event.getSchema());
+            genericDatumWriter.write(event, EncoderFactory.get()
```

Comment:
I think we want to get metadata encoded first
```suggestion
            genericDatumWriter.write(eventMetadata, EncoderFactory.get()
```

Reply:
I'll address it in follow up PR for consumption